### PR TITLE
HDDS-3947: Sort DNs for client when the key is a file for #getFileStatus #listStatus APIs

### PR DESCRIPTION
**Pull request by rakeshadr**
## What changes were proposed in this pull request?

Similar to OzoneManagerFS#lookupFile(OmKeyArgs args, String clientAddress) interface, "clientAddress" has been passed as an argument to OzoneManagerFS#listStatus(), OzoneManagerFS#getFileStatus() APIs.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3947

Please replace this section with the link to the Apache JIRA)

## How was this patch tested?

Used existing TestKeyManagerImpl unit test cases.

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/fs/OzoneManagerFS.java [hunk 1]
```diff
@@ -49,6 +69,21 @@ OpenKeySession createFile(OmKeyArgs args, boolean isOverWrite,
...
+   *                      exists its status is included in the final list.
+   * @param numEntries    Number of entries to list from the start key
+   * @param clientAddress a hint to key manager, order the datanode in returned
+   *                      pipeline by distance between client and datanode.
+   * @return list of file status
+   * @throws IOException if file or bucket or volume does not exist
+   */
   List<OzoneFileStatus> listStatus(OmKeyArgs keyArgs, boolean recursive,
-      String startKey, long numEntries) throws IOException;
+      String startKey, long numEntries, String clientAddress)
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/fs/OzoneManagerFS.java:87]: In my [comment](https://github.com/apache/hadoop-ozone/pull/1192#discussion_r481025587) on the previous PR:

> Should we also keep the two original methods?

I meant both `getFileStatus` and `listStatus`.  Sorry if it was unclear.

**Review by adoroszlai — COMMENTED** [review]
Thanks @rakeshadr for updating the patch.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java [hunk 2]
```diff
@@ -1783,7 +1802,9 @@ private OzoneFileStatus getOzoneFileStatus(String volumeName,
 
       // if the key is a file then do refresh pipeline info in OM by asking SCM
       if (fileKeyInfo != null) {
-        refreshPipeline(fileKeyInfo);
+        if (refreshPipeline) {
```
[the remark this answers is not in the document]
Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java:1805]: Thanks @adoroszlai again for the review help.

Yes, you are correct. Test passed locally without the `if (refreshPipeline)` condition.

Point-1) That means, the caller should call`refreshPipeline(OmKeyInfo value)` method without the `refreshPipeline` condition.

Point-2) Also, I'm planning to move the [refreshPipeline](https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java#L2131) call outside BUCKET_LOCK and will be moving to [here](https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java#L2147). Now, its a duplicate call added as part of HDDS-3658.

Does these two changes make sense to you?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/fs/OzoneManagerFS.java (hunk 1 above)
Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/fs/OzoneManagerFS.java:87]: Sure, I will make the changes in my next commit.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java:1805]: Yes, both points make sense.  Thanks for noticing the duplicate call.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/fs/OzoneManagerFS.java (hunk 1 above)
Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/fs/OzoneManagerFS.java:87]: Sorry for re-opening the discussion on this comment. On a second thought `listStatus` is only used in [one place in the source code](https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java#L2910) and there won't be any other caller for `listStatus() without ClientAddress`. So, how about deleting the original method and can add it later if there is any need ?

Comment (adoroszlai) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/fs/OzoneManagerFS.java:87]: > listStatus is only used in one place in the source code

It is also used by several tests, which all pass `null` as address.

Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/fs/OzoneManagerFS.java:103]: OK, addressed in my new commit

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java]: OK, Done!

**Comment by rakeshadr** [comment]
Thanks @adoroszlai for the reviews. Uploaded new commit addressing your comments.

**Review by xiaoyuyao — APPROVED** [review]
Thanks @rakeshadr  for working on this. The patch LGTM, +1.

**Review by adoroszlai — COMMENTED** [review]
Thanks @rakeshadr for updating the patch.

I have one question left: shouldn't `getFileStatus` call from `listStatus` pass `clientAddress`?

https://github.com/apache/hadoop-ozone/blob/088e40d2f79ab0e8c6021523fd4e2b8db23127af/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java#L2076

**Comment by rakeshadr** [comment]
> Thanks @rakeshadr for updating the patch.
> 
> I have one question left: shouldn't `getFileStatus` call from `listStatus` pass `clientAddress`?
> 
> https://github.com/apache/hadoop-ozone/blob/088e40d2f79ab0e8c6021523fd4e2b8db23127af/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java#L2076

Good catch. Thanks @adoroszlai. 

Updated new commit. I've added clientAddress to the #getFileStatus call. Also, I've moved #getFileStatus call outside BUCKET_LOCK so that it will get the benefit of doing 'sort' and 'refreshPipeline' outside BUCKET_LOCK.

**Comment by rakeshadr** [comment]
> Thanks @rakeshadr for working on this. The patch LGTM, +1.

Thanks a lot @xiaoyuyao for the reviews!

**Review by adoroszlai — APPROVED** [review]
Thanks @rakeshadr for improving this patch several times.

**Comment by adoroszlai** [comment]
Thanks @rakeshadr for the contribution, and @xiaoyuyao for the review.

**Comment by rakeshadr** [comment]
Thanks a lot @adoroszlai for review help and merging it.
